### PR TITLE
Add JSON-marshalable error type

### PR DESCRIPTION
**Pull request by d10v**
Related to https://github.com/facebookincubator/contest/issues/111 and https://github.com/facebookincubator/contest/pull/202#issuecomment-746885326

Very open to suggestions on semantics.

### CODE REVIEW

File: error.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	"errors"
+)
+
+// Error implements a JSON-serializable error
+// "Err" for less confusion over func names
+type Error struct {
+	Err error
+}
+
+func NewError(e error) Error { return Error{Err: e} }
```
Comment (insomniacslk) [inline error.go:14]: Also implement `Unwrap` to support Go1.13+'s error wrapping, see https://jayconrod.com/posts/116/error-handling-guidelines-for-go and https://blog.golang.org/go1.13-errors

File: error.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+func (xe *Error) Error() string { return xe.Err.Error() }
```
Comment (insomniacslk) [inline error.go:16]: is `String` needed?

File: error.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (xe *Error) UnmarshalJSON(b []byte) error {
+	s, err := stripQuotes(string(b))
+	if err != nil {
+		return err
+	}
+	if len(s) == 0 {
+		return errors.New("Error message must not be empty")
```
Comment (insomniacslk) [inline error.go]: nit: in idiomatic Go, error strings should always start with a lower-case letter when possible

**Review by insomniacslk — CHANGES_REQUESTED** [review]
Thanks for your contribution! The code looks good, I've left a few comments inline with some minor requests

File: error.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+func (xe *Error) Error() string { return xe.Err.Error() }
+
+func (xe Error) String() string {
+	return xe.Err.Error()
+}
+
+// MarshalJSON implements the json.Marshaler interface.
+func (xe Error) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + xe.Err.Error() + `"`), nil
```
Comment (tfg13) [inline error.go]:
```suggestion
	return json.Marshal(xe.Err.Error())
```

File: error.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	if err != nil {
+		return err
+	}
+	if len(s) == 0 {
+		return errors.New("Error message must not be empty")
+	}
+	e := errors.New(s)
+
+	*xe = Error{Err: e}
+	return nil
```
Comment (tfg13) [inline error.go]:
```suggestion
	var s string
	err := json.Unmarshal(b, &s)
	if err != nil {
		return err
	}
	xe.Err = errors.New(s)
```

**Review by tfg13 — COMMENTED** [review]
Do we need really need this manual string handling? (real question, I don't know :) )

**Comment by d10v** [comment]
> Do we need really need this manual string handling? (real question, I don't know :) )

I also don't know :) I've just done it like this like in other project places and your way looks cleaner to me.

**Comment by insomniacslk** [comment]
> > Do we need really need this manual string handling? (real question, I don't know :) )
> 
> I also don't know :) I've just done it like this like in other project places and your way looks cleaner to me.

good point - the original (bad) code is mine and this approach is so much better. After this PR I will adjust it in the other places

File: error.go (hunk 1 above)
Comment (d10v) [reply error.go:14]: @insomniacslk love these articles!

**Review by insomniacslk — APPROVED** [review]
Looks good to me! I'll let @tfg13 have a final look before merging. Meanwhile, thank you @d10v for your contribution :)

**Review by tfg13 — APPROVED** [review]

**Comment by d10v** [comment]
Sqash/rebased, and I don't see a button to merge :)
Thanks everyone!

**Comment by insomniacslk** [comment]
Merged! Thanks everyone :)

**Pull request merged**